cmd/main.go

```diff
@@ -2,45 +2,122 @@ package main
 
 import (
 	"context"
+	"errors"
 	"flag"
-	"log"
+	"fmt"
+	"net/http"
 	"os"
 
 	"github.com/joho/godotenv"
 
 	"github.com/benchttp/server"
 	"github.com/benchttp/server/firestore"
+	"github.com/benchttp/server/shutdown"
 )
 
-const defaultPort = "9998"
+const (
+	defaultConfigPath = ".env"
+	defaultPort       = "9998"
+)
 
 func main() {
-	port := flag.String("port", defaultPort, "Address for the server to listen on.")
-	flag.Parse()
-	addr := ":" + *port
+	ctx, cancel := context.WithCancel(context.Background())
+	go shutdown.ListenInterrupt(cancel)
 
-	err := godotenv.Load(".env")
+	shutdownHandle, err := run(ctx, cancel)
 	if err != nil {
-		log.Fatalf("Error loading .env file")
+		shutdownHandle.Call() // nolint
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
 	}
 
-	projectID := os.Getenv("GOOGLE_PROJECT_ID")
-	if projectID == "" {
-		log.Fatalf("GOOGLE_PROJECT_ID variable is not defined")
+	// Wait for interrupt.
+	<-ctx.Done()
+
+	if err := shutdownHandle.Call(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
 	}
+}
+
+// run executes the main program. It starts a server.Server inside
+// a goroutine and returns a configured shutdown.Handle for gracefully
+// shutting down the program.
+//
+// If the bootstrap of the server fails, run returns a non-nil error.
+// Otherwise the caller must use the shutdown.Handle to stop the server.
+func run(ctx context.Context, cancel context.CancelFunc) (*shutdown.Handle, error) {
+	configPath := flag.String("config", defaultConfigPath, "Path to the configuration file (for example --config .env)")
+
+	flag.Parse()
 
-	collectionID := os.Getenv("FIRESTORE_COLLECTION_ID")
-	if projectID == "" {
-		log.Fatalf("FIRESTORE_COLLECTION_ID variable is not defined")
+	config, err := readConfigFile(*configPath)
+	if err != nil {
+		return &shutdown.Handle{}, err
 	}
 
-	rs, err := firestore.NewReportService(context.Background(), projectID, collectionID)
+	rs, err := firestore.NewReportService(context.Background(), config.project, config.collection)
 	if err != nil {
-		log.Fatal(err)
+		return &shutdown.Handle{}, err
+	}
+
+	srv := server.New(config.addr, rs)
+
+	handle := shutdown.NewHandle(func() error {
+		if err := srv.Shutdown(ctx); err != nil && err != context.Canceled {
+			return err
+		}
+
+		if err := rs.Close(); err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	go func() {
+		if err := srv.Start(); err != nil && err != http.ErrServerClosed {
+			fmt.Fprintln(os.Stderr, err)
+			cancel()
+		}
+	}()
+
+	return handle, nil
+}
+
+type config struct {
+	addr       string
+	project    string
+	collection string
```

Comment:
[question] Are collection ids / table names really part of a server config? It's not sensitive nor dependent of the environment to me, so it could be hardcoded as well. (Just wondering, not very important here)